src/mame/drivers/olyboss.cpp

```diff
@@ -0,0 +1,339 @@
+	{
+		int color = BIT(data, 7) ^ rvv;
+		bitmap.pix32(y, (sx * 8) + i) = color?0xffffff:0;
+		data <<= 1;
+	}
+}
+
+/* not used */
+/*
+uint32_t olyboss_state::screen_update_olyboss(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect)
+{
+	uint8_t *chargen = memregion(UPD3301_TAG)->base();
+	//uint8_t *chargen = memregion("chargen")->base();
+	uint8_t y,ra,chr,gfx;
+	uint16_t sy=0,ma=0,x;
+
+	bitmap.fill(0);
+	
+	int numColumns=80;
+	for (y = 28; y > 0; y--)
+	{
+		for (ra = 0; ra < 11; ra++)
+		{
+			uint16_t *p = &bitmap.pix16(sy++);
+
+			for (x = ma; x < ma + numColumns; x++)
+			{
+				chr = m_mainVideoram[x];
+				gfx = chargen[(chr<<4) | ra ];
+				
+				*p++ = BIT(gfx, 7);
+				*p++ = BIT(gfx, 6);
+				*p++ = BIT(gfx, 5);
+				*p++ = BIT(gfx, 4);
+				*p++ = BIT(gfx, 3);
+				*p++ = BIT(gfx, 2);
+				*p++ = BIT(gfx, 1);
+				*p++ = BIT(gfx, 0);
+			}
+		}
+		ma+=numColumns+40;
+	}
+
+
+	return 0;
+}
+*/
+
+WRITE8_MEMBER( olyboss_state::videoram_write )
+{
+	//printf("vramw [%2.2x][%2.2x] port0 [%2.2x] fbfd [%2.2x] fbfe [%2.2x] PC [%4.4x]\n",offset,data,m_port0,m_fbfd,m_fbfe,m_maincpu->safe_pc());
+	m_mainVideoram[offset]=data;
+}
+
+READ8_MEMBER( olyboss_state::videoram_read )
+{
+	return m_mainVideoram[offset];
+}
+
+//**************************************************************************
+//  KEYBOARD
+//**************************************************************************
+
+READ8_MEMBER( olyboss_state::keyboard_read )
+{
+	//printf("keyboard_read offs [%d]\n",offset);
+	if (offset==0)
+	{
+		if (keybHit)
+		{
+			return 0x01;
+		}
+		
+		return 0x00;
+	}
+	else if (offset==1)
+	{
+		if (keybHit)
+		{
+			keybHit=false;
+			return keyStroke;
+		}
+		
+		return 0x00;
+	}
+	
+	return 0x00;
+}
+
+DRIVER_INIT_MEMBER( olyboss_state, olyboss )
+{
+	keybHit=false;
+
+	/* initialize DMA */
+	m_dma->ready_w(1);
+}
+
+void olyboss_state::keyboard_put(u8 data)
+{
+	if (data)
+	{
+		//printf("Keyboard stroke [%2x]\n",data);
+		keyStroke=data;
+		keybHit=true;
+	}
+}
+
+/* 8257 Interface */
+
+WRITE_LINE_MEMBER( olyboss_state::hrq_w )
+{
+	//printf("hrq_w\n");
+	m_maincpu->set_input_line(INPUT_LINE_HALT,state);
+	m_dma->hlda_w(state);
+}
+
+READ8_MEMBER( olyboss_state::dma_mem_r )
+{
+	address_space &program = m_maincpu->space(AS_PROGRAM);
+
+	if (offset==0xf2c6)
+	{
+		printf("DMA read at offset [%x] val is [%2.2x]\n",offset,program.read_byte(offset));
+	}
+	
+	return program.read_byte(offset);
+}
+
+/* ports */
+
+WRITE8_MEMBER( olyboss_state::port_write )
+{
+	printf("Wrote to port [%2.2x] value [%2.2x]\n",offset,data);
+}
+
+READ8_MEMBER( olyboss_state::port_read)
+{
+	printf("Reading from port [%2.2x]\n",offset);
+	return 0xff;
+}
+
+//**************************************************************************
+//  MACHINE CONFIGURATION
+//**************************************************************************
+
+MACHINE_CONFIG_START( olyboss_state::olybossd )
+	MCFG_CPU_ADD(Z80_TAG, Z80, XTAL(4'000'000))
```

Comment:
We're moving towards using the `4_MHz_XTAL` syntactic sugar for this stuff.

Reply:
Hope I caught all your corrections.